### PR TITLE
Allow `IEqualityComparer` in string `(Not)BeEquivalentTo` assertion

### LINKED ISSUE #2339: Allow IEqualityComparer in string equality assertion
### Background and motivation

I'm trying to assert two strings are equal using a custom comparer. The code below works, though it would be nice if the cast to `object` wouldn't be necessary and `Be()` could be used instead of `BeEquivalentTo()`.

```c#
using System.Text.RegularExpressions;
using FluentAssertions;

string logLines = @"one\ntwo\nthree";
string expected = @"one\r\ntwo\r\nthree";

((object)logLines).Should().BeEquivalentTo(expected, options => options.Using(IgnoreLineEndingsComparer.Instance));

public sealed class IgnoreLineEndingsComparer : IEqualityComparer<string>
{
    public static readonly IgnoreLineEndingsComparer Instance = new();

    public bool Equals(string x, string y)
    {
        if (x == y)
        {
            return true;
        }

        if (x == null || y == null)
        {
            return false;
        }

        var xLines = Regex.Split(x, "\r\n|\r|\n");
        var yLines = Regex.Split(x, "\r\n|\r|\n");

        return xLines.SequenceEqual(yLines);
    }

    public int GetHashCode(string obj)
    {
        return obj.GetHashCode();
    }
}
```

### API Proposal

```C#
public class StringAssertions<TAssertions> : ReferenceTypeAssertions<string, TAssertions>
    where TAssertions : StringAssertions<TAssertions>
{
    public AndConstraint<TAssertions> Be(string expected, IEqualityComparer<string> comparer,
        string because = "", params object[] becauseArgs);
}
```

### API Usage

```C#
using System.Text.RegularExpressions;
using FluentAssertions;

string logLines = @"one\ntwo\nthree";
string expected = @"one\r\ntwo\r\nthree";

logLines.Should().Be(expected, IgnoreLineEndingsComparer.Instance);
```

### Alternative Designs

_No response_

### Risks

_No response_

### Are you willing to help with a proof-of-concept (as PR in that or a separate repo) first and as pull-request later on?

No

### PR DESCRIPTION
**Pull request by vbreuss**
Add an overload for `StringAssertions.BeEquivalentTo` and `StringAssertions.NotBeEquivalentTo` that accepts a `IEqualityComparer<string>`.

Details are discussed in #2364.
Fixes #2339.

## IMPORTANT 

* [ ] If the PR touches the public API, the changes have been approved in a separate issue with the "api-approved" label.
* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/develop/Tests/FluentAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [x] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/develop/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://fluentassertions.com/releases).
* [x] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/develop/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/develop/AcceptApiChanges.sh).
* [ ] If the PR affects [the documentation](../tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).
    * [ ] Please also run `./build.sh --target spellcheck` or `.\build.ps1 --target spellcheck` before pushing and check the good outcome

### CODE REVIEW

File: Src/FluentAssertions/Common/StringExtensions.cs [hunk 1]
```diff
@@ -7,6 +8,23 @@ namespace FluentAssertions.Common;
...
+    {
+        for (int index = 0; index < value.Length; index++)
+        {
+            if (index >= expected.Length || !comparer.Equals(value[index..(index + 1)], expected[index..(index + 1)]))
+            {
+                return index;
+            }
+        }
+
+        return -1;
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Common/StringExtensions.cs:25]: 🔧 This line is not covered by a test.

File: Tests/FluentAssertions.Specs/Primitives/StringAssertionSpecs.BeEquivalentTo.cs [hunk 2]
```diff
@@ -12,6 +13,33 @@ public partial class StringAssertionSpecs
 {
     public class BeEquivalentTo
     {
+        [Fact]
+        public void Use_custom_comparer()
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Specs/Primitives/StringAssertionSpecs.BeEquivalentTo.cs:17]: 🤔 I think this test needs a better name

File: Src/FluentAssertions/Primitives/StringAssertions.cs [hunk 3]
```diff
@@ -110,11 +110,40 @@ public AndConstraint<TAssertions> BeOneOf(IEnumerable<string> validValues, strin
     /// <param name="becauseArgs">
     /// Zero or more objects to format using the placeholders in <paramref name="because" />.
     /// </param>
-    public AndConstraint<TAssertions> BeEquivalentTo(string expected, string because = "",
-        params object[] becauseArgs)
+    public AndConstraint<TAssertions> BeEquivalentTo(string expected,
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Primitives/StringAssertions.cs:113]: 🤔 I think we also need to add this to the `NotBeEquivalentTo` API for consistency.

**Review by dennisdoomen — CHANGES_REQUESTED** [review]

File: Tests/FluentAssertions.Specs/Primitives/StringAssertionSpecs.BeEquivalentTo.cs (hunk 2 above)
Comment (vbreuss) [reply Tests/FluentAssertions.Specs/Primitives/StringAssertionSpecs.BeEquivalentTo.cs:17]: I renamed the two tests to `Succeed_for_different_strings_using_custom_matching_comparer` and `Fail_for_same_strings_using_custom_not_matching_comparer`

File: Src/FluentAssertions/Common/StringExtensions.cs (hunk 1 above)
Comment (vbreuss) [reply Src/FluentAssertions/Common/StringExtensions.cs:25]: Really? I changed the value to `-2` and then got 352 failing tests:
image

File: Src/FluentAssertions/Primitives/StringAssertions.cs (hunk 3 above)
Comment (vbreuss) [reply Src/FluentAssertions/Primitives/StringAssertions.cs:113]: I added the additional API changes in #2364 and would implement it in a separate pull request.

File: Src/FluentAssertions/Common/StringExtensions.cs (hunk 1 above)
Comment (vbreuss) [reply Src/FluentAssertions/Common/StringExtensions.cs:25]: Ah, I see. It is line 44 that is no longer covered...
Will check it right away :-)

Comment (vbreuss) [reply Src/FluentAssertions/Common/StringExtensions.cs:25]: I found a way to remove the previous overload of `IndexOfFirstMismatch` to simplify the implementation.

File: Src/FluentAssertions/Primitives/StringAssertions.cs (hunk 3 above)
Comment (IT-VBFK) [reply Src/FluentAssertions/Primitives/StringAssertions.cs:113]: IMHO one PR should consist of both the positive and the negative (if applicable)

File: Tests/FluentAssertions.Specs/Primitives/StringAssertionSpecs.StartWith.cs [hunk 4]
```diff
@@ -26,6 +26,20 @@ public void When_asserting_string_starts_with_the_same_value_it_should_not_throw
...
+        {
+            // Arrange
+            string value = "ABC";
+
+            // Act
+            Action action = () =>
+                value.Should().StartWith(value);
+
+            // Assert
+            action.Should().NotThrow();
```
Comment (IT-VBFK) [inline Tests/FluentAssertions.Specs/Primitives/StringAssertionSpecs.StartWith.cs:40]: We usually call just the assertion now

Comment (vbreuss) [reply Tests/FluentAssertions.Specs/Primitives/StringAssertionSpecs.StartWith.cs:40]: I copied the structure from another test in the same class. Should I adjust all tests or just this one?

Comment (IT-VBFK) [reply Tests/FluentAssertions.Specs/Primitives/StringAssertionSpecs.StartWith.cs:40]: I would say yes :)

File: Src/FluentAssertions/Primitives/StringAssertions.cs (hunk 3 above)
Comment (dennisdoomen) [reply Src/FluentAssertions/Primitives/StringAssertions.cs:113]: Yeah, that's the idea. This also avoids the situation where somebody implements one part and then decides to abandon the the work.

File: Tests/FluentAssertions.Specs/Primitives/StringAssertionSpecs.StartWith.cs (hunk 4 above)
Comment (dennisdoomen) [reply Tests/FluentAssertions.Specs/Primitives/StringAssertionSpecs.StartWith.cs:40]: Let's stick to the boyscout rule. As long as a PR makes things better, it's good enough

File: Src/FluentAssertions/Primitives/StringAssertions.cs (hunk 3 above)
Comment (vbreuss) [reply Src/FluentAssertions/Primitives/StringAssertions.cs:113]: I added the overload for `NotBeEquivalentTo`.

File: Tests/FluentAssertions.Specs/Primitives/StringAssertionSpecs.StartWith.cs (hunk 4 above)
Comment (vbreuss) [reply Tests/FluentAssertions.Specs/Primitives/StringAssertionSpecs.StartWith.cs]: I already adapted the 4 tests in this class :-)

File: docs/_pages/releases.md [hunk 5]
```diff
@@ -13,6 +13,7 @@ sidebar:
 
 ### Improvements
 * Improve failure message for string assertions when checking for equality - [#2307](https://github.com/fluentassertions/fluentassertions/pull/2307)
+* Allow `IEqualityComparer` in string `BeEquivalentTo` assertion - [#2372](https://github.com/fluentassertions/fluentassertions/pull/2372)
```
Comment (dennisdoomen) [inline docs/_pages/releases.md:16]: And `NotBeEquivalentTo`

Comment (vbreuss) [reply docs/_pages/releases.md]: Fixed!
I adapted the PR comment, but totally forgot about the `releases.md` :-)

**Review by dennisdoomen — APPROVED** [review]

**Comment by vbreuss** [comment]
The changes will be implemented with #2413

**Pull request closed without merging**